vxyowsup/whatsapp.py

```diff
@@ -147,6 +151,21 @@ def onMessage(self, messageProtocolEntity):
 
     @ProtocolEntityCallback("receipt")
     def onReceipt(self, entity):
+        '''receives confirmation of delivery to human'''
+        # shady
+        # getting too many receipts
         ack = OutgoingAckProtocolEntity(
             entity.getId(), "receipt", "delivery", entity.getFrom())
         self.toLower(ack)
+
+    @ProtocolEntityCallback("ack")
+    @defer.inlineCallbacks
+    def onAck(self, ack):
+        '''receives confirmation of delivery to server'''
+        # sent_message_id: whatsapp id
+        # user_message_id: vumi_id
+        msg_id = ack.getId()
+        vumi_id = yield self.transport.redis.get(msg_id)
```

Comment:
Redis access should also be from the Twisted thread, so we probably want to add a method on the transport that can be called with the necessary Yowsup information, e.g.: